ultravox/tools/infer_tool.py

```diff
@@ -190,25 +192,46 @@ def dataset_infer(inference: base.VoiceInference, args: InferArgs):
     if args.seed is not None:
         ds_args.shuffle_seed = args.seed
     ds = datasets.create_dataset(args.data_sets[0], ds_args)
-    scores: List[float] = []
-    for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
-        # Store the original question and answer for JSON output.
-        question_text = sample.audio_transcript
-        expected_answer = sample.messages[-1]["content"]
-        # Drop any assistant response from the sample.
-        sample.messages = sample.messages[:-1]
-        if not args.json:
-            run_tui(i, inference, sample, args, expected_answer, scores)
-        else:
-            output = inference.infer(
-                sample, max_tokens=args.max_tokens, temperature=args.temperature
+
+    if args.json:
+        dl = data_utils.DataLoader(
+            datasets.Range(ds, args.num_samples),
+            batch_size=args.batch_size,
+            collate_fn=lambda x: x,
+        )
+        current_batch = []
+        sample_index = 0
+        for current_batch in dl:
+            output = []
+            for sample in current_batch:
+                output.append(
+                    {
+                        "index": sample_index,
+                        "question": sample.audio_transcript,
+                        "expected_answer": sample.messages[-1]["content"],
+                    }
+                )
+                sample.messages = sample.messages[:-1]
+                sample_index += 1
```

Comment:
A total nit, but I felt like this would be a bit cleaner (see next comment for related change):
```suggestion
            expected_answers = [sample.messages.pop()["content"] for sample in current_batch]
```